src/consumer/consumerGroup.js

```diff
@@ -272,6 +283,44 @@ module.exports = class ConsumerGroup {
     })
   }
 
+  joinAndSync() {
+    const startJoin = Date.now()
+    return this.retrier(async bail => {
+      try {
+        await this[PRIVATE.JOIN]()
+        await this[PRIVATE.SYNC]()
+
+        const memberAssignment = this.assigned().reduce(
+          (result, { topic, partitions }) => ({ ...result, [topic]: partitions }),
+          {}
+        )
+
+        const payload = {
+          groupId: this.groupId,
+          memberId: this.memberId,
+          leaderId: this.leaderId,
+          isLeader: this.isLeader(),
+          memberAssignment,
+          groupProtocol: this.groupProtocol,
+          duration: Date.now() - startJoin,
+        }
+
+        this.instrumentationEmitter.emit(GROUP_JOIN, payload)
+        this.logger.info('Consumer has joined the group', payload)
+      } catch (e) {
+        if (isRebalancing(e)) {
+          // Rebalance in progress isn't a retriable error since the consumer
+          // has to go through find coordinator and join again before it can
+          // actually retry. Throwing a retriable error to allow the retrier
+          // to keep going
+          throw new KafkaJSError('The group is rebalancing')
```

Comment:
I'm having trouble parsing the comment: "It isn't retriable [...] therefore throw a retriable error".

Reply:
Heh, yeah, I had the same issue. Took me a bit to figure out what it meant (I just moved it from the runner). Maybe we can come up with a better phrasing.

What it's saying is that the errors that we consider "rebalancing errors" are not retriable errors according to the protocol, but they are operations that we want to handle and retry - so we want the retrier to actually retry, not just bail out. Therefore we turn them into retriable errors even though they technically aren't.

Reply:
> While "rebalance in progress" isn't a retriable error in the protocol [since the consumer ...], we can throw a retriable `KafkaJSError` here to restart that whole sequence using the retrier itself.

Something like that?

Reply:
a051f0f hopefully improved it a bit.

Reply:
I understood that last one first try.